packages/chain-bridge/src/router/BridgeRouter.sol

```diff
@@ -0,0 +1,828 @@
+                        sourceChainId
+                    )
+                );
+                delivered = success;
+            }
+        } catch {
+            delivered = false;
+        }
+
+        // Emit event based on delivery result
+        emit ReadResponseDelivered(operationId, originator, delivered);
+
+        // Only update status if delivery failed
+        if (!delivered) {
+            operationStatuses[operationId] = BridgeTypes.OperationStatus.FAILED;
+            emit OperationStatusUpdated(
+                operationId,
+                BridgeTypes.OperationStatus.FAILED
+            );
+        }
+    }
+
+    /*//////////////////////////////////////////////////////////////
+                            VIEW FUNCTIONS
+    //////////////////////////////////////////////////////////////*/
+
+    /// @inheritdoc IBridgeRouter
+    function getAdapters() public view returns (address[] memory) {
+        return adapters.values();
+    }
+
+    /// @inheritdoc IBridgeRouter
+    function isValidAdapter(address adapter) external view returns (bool) {
+        return adapters.contains(adapter);
+    }
+
+    /// @inheritdoc IBridgeRouter
+    function getOperationStatus(
+        bytes32 operationId
+    ) external view returns (BridgeTypes.OperationStatus) {
+        return operationStatuses[operationId];
+    }
+
+    /*//////////////////////////////////////////////////////////////
+                           ADMIN FUNCTIONS
+    //////////////////////////////////////////////////////////////*/
+
+    /// @inheritdoc IBridgeRouter
+    function registerAdapter(address adapter) external onlyGovernor {
+        if (adapters.contains(adapter)) revert AdapterAlreadyRegistered();
+
+        adapters.add(adapter);
+        emit AdapterRegistered(adapter);
+    }
+
+    /// @inheritdoc IBridgeRouter
+    function removeAdapter(address adapter) external onlyGovernor {
+        if (!adapters.contains(adapter)) revert UnknownAdapter();
+
+        adapters.remove(adapter);
+        emit AdapterRemoved(adapter);
+    }
+
+    /// @inheritdoc IBridgeRouter
+    function pause() external onlyGuardianOrGovernor {
+        paused = true;
+    }
+
+    /// @inheritdoc IBridgeRouter
+    function unpause() external onlyGovernor {
+        paused = false;
+    }
+
+    /// @inheritdoc IBridgeRouter
+    function recoverFunds(
+        address recipient,
+        uint256 amount
+    ) external onlyGovernor nonReentrant {
+        if (recipient == address(0)) revert InvalidParams();
+        if (address(this).balance < amount) revert InsufficientBalance();
+
+        (bool success, ) = recipient.call{value: amount}("");
+        if (!success) revert TransferFailed();
+
+        emit RouterFundsRecovered(recipient, amount);
+    }
+
+    /// @inheritdoc IBridgeRouter
+    function setChainRouterAddress(
+        uint16 chainId,
+        address routerAddress
+    ) external onlyGovernor {
+        chainToRouterAddress[chainId] = routerAddress;
```

Comment:
this should be part of the registry as single source of truth - so we have one to many realtion between the bridge routers